File: banquet-pilot/src/core/loader.py
"""Load level JSON files."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict

LEVELS_DIR = Path(__file__).resolve().parents[2] / "levels"


def load_level(level_id: str) -> Dict[str, Any]:
    name = level_id.lower()
    path = LEVELS_DIR / f"{name}.json"
    if not path.exists():
        # allow L01 -> l01
        path = LEVELS_DIR / f"{name.replace('L', 'l')}.json"
    with path.open(encoding="utf-8") as f:
        return json.load(f)


def load_all_levels() -> Dict[str, Dict[str, Any]]:
    out: Dict[str, Dict[str, Any]] = {}
    for path in sorted(LEVELS_DIR.glob("l*.json")):
        with path.open(encoding="utf-8") as f:
            data = json.load(f)
        out[data["id"]] = data
    return out

File: banquet-pilot/src/core/loader.js
/** Load level JSON files (Node). Browser code should import JSON or fetch instead. */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { validateLevel } from "./rules.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
export const LEVELS_DIR = path.resolve(__dirname, "../../levels");

/**
 * @param {string} levelId
 */
export function loadLevel(levelId) {
  const name = String(levelId).toLowerCase();
  let filePath = path.join(LEVELS_DIR, `${name}.json`);
  if (!fs.existsSync(filePath)) {
    filePath = path.join(LEVELS_DIR, `${name.replace(/L/g, "l")}.json`);
  }
  const raw = fs.readFileSync(filePath, "utf8");
  const data = JSON.parse(raw);
  const errors = validateLevel(data);
  if (errors.length > 0) {
    throw new Error(`invalid level ${data.id || levelId}: ${errors.join("; ")}`);
  }
  return data;
}

export function loadAllLevels() {
  /** @type {Record<string, object>} */
  const out = {};
  const files = fs
    .readdirSync(LEVELS_DIR)
    .filter((f) => /^l.*\.json$/i.test(f))
    .sort();
  for (const f of files) {
    const data = JSON.parse(fs.readFileSync(path.join(LEVELS_DIR, f), "utf8"));
    const errors = validateLevel(data);
    if (errors.length > 0) {
      throw new Error(`invalid level ${data.id}: ${errors.join("; ")}`);
    }
    out[data.id] = data;
  }
  return out;
}
